wp-includes/js/service-worker.js

```diff
@@ -0,0 +1,52 @@
+/* global workbox */
+
+/**
+ * Handle registering caching strategies.
+ */
+
+// @todo Perhaps some other variable makes more sense?
+var wp = {};
+
+wp.serviceWorker = {
+
+	/**
+	 * Add caching strategy. Missing processing conflicting routes.
+	 *
+	 * @param {string}      route Route.
+	 * @param {string}      strategy Strategy.
+	 * @param {string|null} cacheName Cache name.
+	 * @param {int}         maxAge Max Age.
+	 * @param {int}         maxEntries Max entries.
+	 */
+	addCachingStrategy: function ( route, strategy, cacheName, maxAge, maxEntries ) {
```

Comment:
I wonder if this abstraction would be necessary or if we'd be largely reinventing the wheel that Workbox already provides. In other words, what if `wp.serviceWorker` were just an alias to `workbox`? So then instead of calling `workbox.routing.registerRoute()` we could instead call `wp.serviceWorker.routing.registerRoute()`?

Reply:
Could be. Added this abstraction mainly thinking of having all the direct usage of Workbox in one JS file and to only use the abstraction in PHP files in case for some reason core would move away from Workbox. Then just this JS file would need to be modified to replace the logic while keeping the abstraction method the same. However, I might be overthinking, and it's true that using abstraction like this might be redundant. In this case we could for example remove the `service-worker.js` file and define `wp.serviceWorker` within `WP_Service_Workers::get_workbox_script()`. And then use `wp.serviceWorker.routing.registerRoute()` within `WP_Service_Workers::register_caching_strategy_for_route()` instead. Thoughts?

Reply:
Removing the abstraction layer to use `wp.serviceWorker` as an alias to `workbox` for now.

Reply:
Just a quick heads-up in case anything catches you out: the `workbox.*` namespace is, for the most part, backed by a [`Proxy`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy) that will [dynamically load](https://github.com/GoogleChrome/workbox/blob/096e44897e9c348f85bb51927c1113de4e339897/packages/workbox-sw/controllers/WorkboxSW.mjs#L59-L72) (via `importScripts()`) the underlying `.js` on demand.

One wrinkle is that calls to `importScripts()` need to happen as part of the initial, synchronous execution of a service worker script—as opposed to, e.g., inside of a callback registered with `addEventListener()`.

That's all to say: there's a chance that you might run into unexpected errors that are hard to diagnose if you end up making the initial call to, e.g., a `wp.serviceWorker.strategies.*` method inside of, e.g., a custom `fetch` event listener. I wanted to explicitly call this out to head off any really confusing debugging 😄 

(There's more context, including some workarounds for synchronously preloading the required scripts, at https://github.com/GoogleChrome/workbox/issues/1504)

Reply:
@jeffposnick thank you for that helpful info.

In the PR now there is a `wp.serviceWorker = workbox` statement to achieve an alias. However, the `workbox` name still remains and it is important to remain currently because the imported scripts all extend that global object from what I understand. 

To recap, we discussed creating a wrapper API around Workbox so that in the future we could potentially swap out Workbox with another implementation later, by however unlikely chance we’d need to. This was done in WordPress with the introduction of CodeMirror by creating a `wp.codeEditor` abstraction. Now with the Gutenberg there is the `wp.element` namespace which exports the React functionality; if WordPress moves away from React in the future, then `wp.element.createElement` could be easily swapped out with another implementation that has the same behaviors. So we’d like to do something similar with Workbox in the service worker, perhaps making Workbox available as `wp.serviceWorker` as seen here now.

All of this to say, it doesn’t seem that the `workbox copyLibraries` command makes this possible to use a different export namespace than the `workbox.*` global object. Is that something which would need to be added, or is it possible to do now using an alternative means?

Reply:
We don't allow for customizing the top-level namespace when using `workbox copyLibraries`, but `wp.serviceWorker = workbox` seems like a reasonable approach to indirection.

We do ship `.mjs` files (i.e. JavaScript modules) with explicit exports of each of the underlying classes/static methods, without a namespace. It's possible to create a source file that references only the exports you need, pulling them all into a distinct namespace that you control, and then running that through `webpack` or `Rollup` to create the final bundle. That's really the alternative that we offer for folks who need more than our default `workbox.*`-based dynamic loading experience.

If you're happy with the `wp.serviceWorker = workbox` approach, especially when prototyping, I don't think there's any need to switch.

Reply:
The only concern with the `wp.serviceWorker = workbox` approach is that third party code could start to directly access `workbox`, which we'd want to avoid.

Also, since we can't anticipate the functionality that third party code will leverage we need to have the dynamic imports, instead of creating a build with Webpack.

What we have is fine for now, but once we get to the point of proposing for WordPress core merge then this issue will come up again.